File: packages/integrations/examples/vercel/src/smoke.ts
import { strict as assert } from "node:assert";
import { fileURLToPath } from "node:url";

import { createMCPClient, type MCPClient } from "@ai-sdk/mcp";
import { Experimental_StdioMCPTransport } from "@ai-sdk/mcp/mcp-stdio";
import type { Tool } from "ai";

const stdioServerPath = fileURLToPath(
  new URL("../../../dist/codemode/stdio-server.mjs", import.meta.url),
);
let client: MCPClient | undefined;

try {
  client = await createMCPClient({
    transport: new Experimental_StdioMCPTransport({
      command: process.execPath,
      args: [stdioServerPath],
      env: localSmokeEnvironment(),
    }),
  });
  const tools = await client.tools();
  assert.deepEqual(Object.keys(tools), ["code_execute"]);

  const tool = tools.code_execute as Tool<{ code: string }, unknown>;
  assert.equal(typeof tool?.execute, "function");
  const first = await tool.execute!(
    {
      code: `
        await page.goto("https://example.com", { waitUntil: "load" });
        await context.newPage();
        await context.setActivePage(page);
        return { title: await page.title(), pages: (await context.pages()).length };
      `,
    },
    { context: {}, messages: [], toolCallId: "vercel-smoke-1" },
  );
  const second = await tool.execute!(
    { code: `return { title: await page.title(), pages: (await context.pages()).length };` },
    { context: {}, messages: [], toolCallId: "vercel-smoke-2" },
  );

  assert.ok(containsBrowserState(first, "Example Domain", 2), JSON.stringify(first));
  assert.ok(containsBrowserState(second, "Example Domain", 2), JSON.stringify(second));
  process.stdout.write(
    `${JSON.stringify({ status: "PASS", tools: ["code_execute"], statePersisted: true })}\n`,
  );
} finally {
  await client?.close().catch(() => undefined);
}

function localSmokeEnvironment(): Record<string, string> {
  const environment: Record<string, string> = { STAGEHAND_BROWSER: "local" };
  for (const name of ["CHROME_PATH", "HOME", "PATH", "TMPDIR"]) {
    const value = process.env[name];
    if (value) environment[name] = value;
  }
  return environment;
}

function containsBrowserState(value: unknown, title: string, pages: number): boolean {
  if (Array.isArray(value)) return value.some((entry) => containsBrowserState(entry, title, pages));
  if (typeof value !== "object" || value === null) return false;
  const record = value as Record<string, unknown>;
  if (record.title === title && record.pages === pages) return true;
  return Object.values(record).some((entry) => containsBrowserState(entry, title, pages));
}

File: packages/integrations/examples/crewai/smoke.py
from __future__ import annotations

import json
import os
from collections.abc import Iterator
from contextlib import contextmanager
from pathlib import Path
from typing import Any

from crewai.tools import BaseTool
from crewai_tools import MCPServerAdapter

from agent import (
    REPOSITORY_ROOT,
    build_stagehand_agent,
    load_stagehand_codemode_skill,
)
from mcp import StdioServerParameters

LOCAL_STDIO_SERVER_PATH = REPOSITORY_ROOT / "packages/integrations/dist/codemode/stdio-server.mjs"


@contextmanager
def local_stagehand_code_tools_for_ci() -> Iterator[list[BaseTool]]:
    """Launch the MCP directly only for the credential-free local CI smoke."""
    if not LOCAL_STDIO_SERVER_PATH.is_file():
        raise FileNotFoundError(
            "Build @browserbasehq/stagehand-integrations before running the smoke"
        )
    child_env = {
        "PATH": os.environ.get("PATH", ""),
        "STAGEHAND_BROWSER": "local",
    }
    # Stagehand's local launcher uses CI to add Chrome's --no-sandbox flag.
    for name in ("HOME", "TMPDIR", "CHROME_PATH", "CI"):
        if value := os.environ.get(name):
            child_env[name] = value
    parameters = StdioServerParameters(
        command="node",
        args=[str(LOCAL_STDIO_SERVER_PATH)],
        cwd=Path(REPOSITORY_ROOT),
        env=child_env,
    )
    with MCPServerAdapter(parameters) as discovered_tools:
        yield list(discovered_tools)


def successful_result(raw_result: Any) -> dict[str, Any]:
    result = json.loads(str(raw_result))
    assert result["ok"] is True, result
    return result


def main() -> None:
    os.environ.setdefault("OPENAI_API_KEY", "smoke-only-placeholder")

    with local_stagehand_code_tools_for_ci() as tools:
        assert [tool.name for tool in tools] == ["code_execute"]
        assert "Stagehand V4 code-mode syntax" in tools[0].description
        assert "stagehand.extract" in load_stagehand_codemode_skill()

        agent = build_stagehand_agent(tools)
        code_execute = agent.tools[0]

        first = successful_result(
            code_execute.run(
                code="""
await page.goto("https://example.com", { waitUntil: "domcontentloaded" });
await page.evaluate(() => {
  document.documentElement.dataset.crewaiStagehandSession = "persisted";
});
return {
  pageId: page.pageId,
  title: await page.title(),
  marker: await page.evaluate(
    () => document.documentElement.dataset.crewaiStagehandSession,
  ),
};
"""
            )
        )
        second = successful_result(
            code_execute.run(
                code="""
return {
  pageId: page.pageId,
  title: await page.title(),
  marker: await page.evaluate(
    () => document.documentElement.dataset.crewaiStagehandSession,
  ),
};
"""
            )
        )

        first_value = first["value"]
        second_value = second["value"]
        assert first_value["title"] == "Example Domain"
        assert second_value["title"] == "Example Domain"
        assert second_value["marker"] == "persisted"
        assert second_value["pageId"] == first_value["pageId"]

        print(
            "CrewAI local CI-only Stagehand MCP persistence PASS:",
            json.dumps(
                {
                    "browser": "local",
                    "tool": "code_execute",
                    "pageId": second_value["pageId"],
                    "title": second_value["title"],
                    "marker": second_value["marker"],
                },
                sort_keys=True,
            ),
        )


if __name__ == "__main__":
    main()
